Src/PerformanceCollector/Shared/Implementation/QuickPulse/Process/PerfLib/CategorySample.cs

```diff
@@ -0,0 +1,258 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.Implementation.QuickPulse.PerfLib
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Runtime.InteropServices;
+
+    internal class CategorySample
+    {
+        public Dictionary<int, CounterDefinitionSample> CounterTable;
+
+        public Dictionary<string, int> InstanceNameTable;
+
+        private readonly PerfLib library;
+
+        public CategorySample(byte[] data, int categoryNameIndex, int counterNameIndex, PerfLib library)
+        {
+            if (library == null)
+            {
+                return;
+            }
+
+            this.library = library;
+
+            NativeMethods.PERF_DATA_BLOCK dataBlock = new NativeMethods.PERF_DATA_BLOCK();
+
+            GCHandle handle = GCHandle.Alloc(data, GCHandleType.Pinned);
+
+            try
+            {
+                IntPtr dataRef = handle.AddrOfPinnedObject();
+
+                Marshal.PtrToStructure(dataRef, dataBlock);
+
+                dataRef = (IntPtr)((long)dataRef + dataBlock.HeaderLength);
+
+                int numPerfObjects = dataBlock.NumObjectTypes;
+                if (numPerfObjects == 0)
+                {
+                    this.CounterTable = new Dictionary<int, CounterDefinitionSample>();
+                    this.InstanceNameTable = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+                    return;
+                }
+
+                // Need to find the right category, GetPerformanceData might return several of them.
+                NativeMethods.PERF_OBJECT_TYPE perfObject = new NativeMethods.PERF_OBJECT_TYPE();
```

Comment:
We're using the "pseudo-registry" perf interface, which is used internally by both PDH and CLR's PerformanceCounter. The minimum amount of perf information that can be read is a single performance object (aka category) - in our case "Process". It does include all counters, but only within the object; the same thing happens when you think you read a single performance counter in CLR, for instance. No custom perf provider will have any role to play here, since it's all just the "Process" object.